Bug 1210304 - HomeScreen for RedSquare has been implemented

TatianaMigacheva opened the pull request:
This is delivery of HomeScreen feature for RedSquare project. 
Meta Bug 1210304 contains the following bugs:
Bug 1218283 - [RedSquare] Split homescreen on two views
Bug 1218325 - [RedSquare] Homescreen - add dynamic list of elements in "pin apps" screen
Bug 1215097 - [RedSquare] Implement touchless navigation for homescreen's pin apps
Bug 1217816 - [RedSquare] HomeScreen application More Apps view
Bug 1216931 - [RedSquare] Add clock on main screen

ruvsmirnov committed:
Bug 1218283 - [RedSquare] Split homescreen on two views

ruvsmirnov committed:
Bug 1215097 - It was implemented vertical navigation by pin up homescreen

MariaKolesina committed:
Bug 1218325 - dynamic list of elements has been added to main-screen

Pin apps list with predetermined number of elements changed to dynamic list

ruvsmirnov committed:
Bug 1216931 - Add clock on home screen.

autolander commented:
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=06a972ade922cb6b51a8a37721cbb21201f7f80e

davidflanagan commented on apps/settings/elements/homescreen.html:
```diff
@@ -13,12 +13,6 @@ <h2 data-l10n-id="vertical-homescreen-layout">Layout</h2>
-                <option value="3" data-l10n-id="vertical-homescreen-three-columns">Three Columns</option>
-                <option value="4" data-l10n-id="vertical-homescreen-four-columns">Four Columns</option>
-              </select>
-            </div>
```
Just removing this one select element leave us with an empty panel, doesn't it? I think you should probably remove this entire file and the references to it, right?

davidflanagan replied:
And of course you should also remove the homescreen code that reads this grid.layout.cols setting, whereever that code is.

davidflanagan commented on apps/verticalhome/build/default-homescreens.json:
```diff
@@ -24,6 +24,6 @@
   ],
   "preferences": {
-    "grid.cols": 3
+    "grid.cols": 2
```
Given that the number of columns will not be configurable, I think that it doesn't need to be in this JSON file anymore.  Maybe find the code that reads this file and just define a const for the number of columns.

davidflanagan commented on apps/verticalhome/index.html:
```diff
@@ -68,50 +73,56 @@
-          <button id="edit-group-title-clear" type="reset"></button>
-        </p>
-      </form>
+    <section id="more-apps-screen" role="region" hidden="">
```
just use the hidden attribute with no value. No need for the =""

davidflanagan commented on apps/verticalhome/index.html:
```diff
@@ -68,50 +73,56 @@
-      </form>
+    <section id="more-apps-screen" role="region" hidden="">
+      <div class='more-apps-header'>More Apps</div>
+      <gaia-grid-rs id="icons"></gaia-grid-rs>
```
Since you're only using the gaia-grid for the more apps, you can dramatically improve the homescreen startup performance by not loading all of the gaia-grid code until it is actually needed.

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -126,10 +132,19 @@
       }.bind(this));
+
+      this.pinManager = new PinAppManager();
+      navigator.mozL10n.ready(this.l10nInit.bind(this));
```
Since this is all related to the time format, it would be better to put it in the clock module.

Also note that .ready() only gets called once. You probably want to be notified every time the locale changes, so that the clock format can change dynamically if the user switches locales.

ruvsmirnov replied:
We add event listener for timeformatchange event where we refresh time format and locale also.
I check this case. Localization for clock on homescreen changes.

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -126,10 +132,19 @@
+
+        this.pinNavigation = new PinAppNavigation(document.getElementById('pin-apps-list'));
+        this.pinNavigation.points_selector = '#pin-apps-list .pin-app-item';
+      }.bind(this));
```
I don't get what any of this code is doing. It is my first time reading through it, but this seems like a place where comments would be helpful.

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -141,6 +156,38 @@
+      this.inMoreApps = true;
+      console.log("inMoreApps is : " + this.inMoreApps);
+      document.getElementById("main-screen").setAttribute("hidden","");
+      document.getElementById("more-apps-screen").removeAttribute("hidden");
```
Rather than using the hidden attribute here, it is probably better to set and remove a CSS class that will show the more apps screen on top of the main screen. That way you can specify CSS animations for the transition, which is what our UX designers have in mind, I think.

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -141,6 +156,38 @@
+      document.getElementById("main-screen").setAttribute("hidden","");
+      document.getElementById("more-apps-screen").removeAttribute("hidden");
+      MoreAppsNavigation.init();
+      document.getElementsByTagName('gaia-grid-rs')[0].scrollTo(0,0);
```
this line seems like someting that ought to be in the init() call.

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -141,6 +156,38 @@
+    },
+
+    backToMainScreen: function() {
+      this.__proto__.inMoreApps = false;
```
If you ever have to use __proto__ your code is almost certainly wrong. Accessing __proto__ is also a big performance hit, since the code can't be optimized by the JS interpreter.  

You'll have to have a very good reason for __proto__ and a long comment explaining the reason, or I will not land the patch.

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -141,6 +156,38 @@
+      document.getElementsByTagName('gaia-grid-rs')[0].scrollTo(0,0);
+    },
+
+    backToMainScreen: function() {
```
To be parallel with the function above, I'd suggest you call this hideMoreApps()

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -157,14 +204,43 @@
       this.renderGrid();
```
I'm guessing that this onLocalized() function will need to be updated so it doesn't just localize and render the more apps grid, but also the list of pinned apps.

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -157,14 +204,43 @@
+
+    refreshClock: function ls_refreshClock(now) {
+      var f = new navigator.mozL10n.DateTimeFormat();
+      var _ = navigator.mozL10n.get;
```
It seems silly to do this for just once use.

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -157,14 +204,43 @@
+
+      var timeFormat = this.timeFormat.replace('%p', '<span>%p</span>');
+      var dateFormat = _('longDateFormat');
+      this.clockTime.innerHTML = f.localeFormat(now, timeFormat);
```
writing to innerHTML will cause trouble on a security audit. It is okay to set innerHTML to the empty string to delete elements, but setting this to a string, especially something that is not a string literal is considered a risk of bugs. You'll have to change this before we can land it.

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -157,14 +204,43 @@
+      var _ = navigator.mozL10n.get;
+
+      var timeFormat = this.timeFormat.replace('%p', '<span>%p</span>');
+      var dateFormat = _('longDateFormat');
```
It doesn't seem right to do localization over and over again every minute. That's inefficient. Please separate the code for updating the formats from the code for updating the locales.

And if you can, move this code into your clock.js module.

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -327,7 +403,35 @@
         // A hashchange event means that the home button was pressed.
         // The system app changes the hash of the homescreen iframe when it
         // receives a home button press.
+        case 'timeformatchange':
```
You inserted this new case right after a comment that is supposed to go with the next case.  Please fix that.  It would also be good to add a comment explaining where the timeformatchagne event comes from, since it does not appear anywhere else in this app directory.

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -327,7 +403,35 @@
         case 'hashchange':
+          this.backToMainScreen();
+          this.pinNavigation.reset();
+          this.clockTime.parentNode.style.opacity = 1;
```
Setting an individual style like this seems wrong, and having to use parentNode seems wrong.  I'd define a variable to hold the parent of the clockTime element.  And I think you should define a CSS class and use that rather than hardcoding the opacity here.

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -327,7 +403,35 @@
+          this.timeFormat = window.navigator.mozHour12 ?
+            navigator.mozL10n.get('shortTimeFormat12') :
+            navigator.mozL10n.get('shortTimeFormat24');
+          this.refreshClock(new Date());
```
It doesn't seem right that you'd have to do this here on the home button.

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -327,7 +403,35 @@
+          this.refreshClock(new Date());
+
+
+
```
too many blank lines

davidflanagan commented on apps/verticalhome/js/app.js:
```diff
@@ -327,7 +403,35 @@
+
+
           // The group editor UI will be hidden by itself so returning...
           var editor = exports.groupEditor;
```
I don't really know what the group editor UI is, but it sounds like code that could be removed.

davidflanagan commented on apps/verticalhome/js/app_manager.js:
```diff
@@ -7,7 +7,7 @@
 
   function AppManager() {
-    var grid = document.querySelector('gaia-grid');
+    var grid = document.querySelector('gaia-grid-rs');
```
If I understand correctly, the gaia-grid is now only used for the more apps section, right? So maybe this entire module should be renamed to MoreAppsManager or something.  (If you do a git mv you can rename a file without making the entire file look like new code)

davidflanagan commented on apps/verticalhome/js/clock.js:
```diff
@@ -0,0 +1,86 @@
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
```
We don't have these comments at the top of the other new files, so let's not start adding them now.  The fact that the files are being submitted as pull requests to a repo with a LICENSE file in it will have to be good enough to indicate the license

davidflanagan commented on apps/verticalhome/js/clock.js:
```diff
@@ -0,0 +1,86 @@
+  limitations under the License.
+*/
+
+'use strict';
```
I think it is more useful to put the 'use strict' inside the '(function(exports) {'.  When this code ships, all the files will be concatenated. So most 'use strict' directives like this one will not be the first expression in the file and will have no effect.  If you put them as the first expression inside a function then they will always have an effect.

davidflanagan commented on apps/verticalhome/js/clock.js:
```diff
@@ -0,0 +1,86 @@
+
+    if (this.timeoutID == null) {
+      this.timeoutID = window.setTimeout(function cl_setClockInterval() {
+
```
you should probaby set self.timeoutID to null here

davidflanagan commented on apps/verticalhome/js/clock.js:
```diff
@@ -0,0 +1,86 @@
+    var date = new Date();
+    var self = this;
+
+    refresh(date);
```
This would be a little bit cleaner and decoupled if you did window.dispatchEvent(new CustomEvent('tick')) here instead of calling the refresh callback.  I don't feel strongly about that, though.

davidflanagan commented on apps/verticalhome/js/clock.js:
```diff
@@ -0,0 +1,86 @@
+      window.clearInterval(this.timerID);
+      this.timerID = null;
+    }
+  };
```
Somewhere, you should have code that stops the clock when the homescreen is hidden by another app and restarts the clock when the homescreen becomes visible again.  I saw you have that for more apps, but you should have it for when other apps are launched, too.  We need to be very certain that the timers stop running when the screen is blanked or the CPUs will never sleep fully.

You need to listen for mozvisibilitychange events, I think.  Putting it in this module might be the cleanest way to do it.

davidflanagan replied:
Oops. No prefix, just "visibilitychange" events. There is already code that does that, but not related to the clock, I think.

davidflanagan commented on apps/verticalhome/js/configurator.js:
```diff
@@ -188,8 +200,15 @@
 
+  function loadPinApps() {
+    LazyLoader.getJSON(DEFAULT_PIN_APP_CONF_FILE).then(
+      onLoadPinAppsJSON, onErrorPinApps
```
why not move the definition of these two functions closer.

davidflanagan commented on apps/verticalhome/js/configurator.js:
```diff
@@ -188,8 +200,15 @@
+
   function Configurator() {
     load();
+    loadPinApps();
```
I'm not sure what the configurator does. Is this code that only runs the very first time after flashing the phone?  I assume that after the first run the list of pinned apps will be read from some kind of local storage, right?

davidflanagan commented on apps/verticalhome/js/more_apps_navigation.js:
```diff
@@ -0,0 +1,95 @@
+          this.changeSelectElem(deltaIndex);
+          if (!this.isVisible(this.selectElement)) {
+            this.selectElement.childNodes[0].scrollIntoView({behavior:"smooth", block: "end"});
+          }
```
There is a lot of duplicated code for these four arrow keys. Can you write a utility function that just takes the deltaIndex and does the right thing?  I'd also like to see a comment that explains why you sometimes have to use childNodes[0] and sometimes do not have to.

davidflanagan commented on apps/verticalhome/js/more_apps_navigation.js:
```diff
@@ -0,0 +1,95 @@
+
+    handleEvent: function (e) {
+      var deltaIndex = 0;
+      e.preventDefault();
```
don't call preventDefault until you know you're going to actually handle the keys. This line breaks the volume keys, for example.  You'll need to call preventDefault in each of the individual cases below.

davidflanagan commented on apps/verticalhome/js/more_apps_navigation.js:
```diff
@@ -0,0 +1,95 @@
+          break;
+        case 'Accept':
+          window.removeEventListener('keydown', this);
+          this.selectElement.click();
```
Does this launch the app? That is worth mentioning in a comment. Should you call backToMainScreen() here?

davidflanagan commented on apps/verticalhome/js/more_apps_navigation.js:
```diff
@@ -0,0 +1,95 @@
+    selectElement: null,
+
+    init: function () {
+      this.controls = app.grid.getItems();
```
Why "controls" rather than apps or icons or items?

davidflanagan commented on apps/verticalhome/js/more_apps_navigation.js:
```diff
@@ -0,0 +1,95 @@
+
+    init: function () {
+      this.controls = app.grid.getItems();
+      this.reset();
```
Strange to call reset() after setting the controls. I have to go check whether that resets the controls array.

davidflanagan commented on apps/verticalhome/js/more_apps_navigation.js:
```diff
@@ -0,0 +1,95 @@
+      this.reset();
+    },
+
+    reset: function () {
```
Does anything other than init() call reset? Could you just move this code to init?

davidflanagan commented on apps/verticalhome/js/more_apps_navigation.js:
```diff
@@ -0,0 +1,95 @@
+      this.selectElement = this.controls[this.selectedElemIndex].element;
+      this.selectElement.classList.add('selected');
+      this.clearNavigationEvents();
+      window.addEventListener('keydown', this);
```
This should be in init(), not reset(), I think.  When you close the more apps, I assume that you're calling reset, right? So you don't want an active event listener in that case, right?

davidflanagan commented on apps/verticalhome/js/more_apps_navigation.js:
```diff
@@ -0,0 +1,95 @@
+
+    clearNavigationEvents: function () {
+      window.removeEventListener('keydown', this);
+    },
```
I don't think you need a special function for this, do you?

davidflanagan commented on apps/verticalhome/js/more_apps_navigation.js:
```diff
@@ -0,0 +1,95 @@
+        case 'ArrowUp':
+          deltaIndex = -2;
+          this.changeSelectElem(deltaIndex);
+          if (!this.isVisible(this.selectElement)) {
```
Do you need the isVisible() check? Won't scrollIntoView do the right thing if the icon is already visible?

davidflanagan commented on apps/verticalhome/js/more_apps_navigation.js:
```diff
@@ -0,0 +1,95 @@
+      this.selectElement.classList.add('selected');
+    },
+
+    isVisible: function (element) {
```
Do you really need this?  If so, dosn't the grid component already do it for you somehow?  This doesn't seem like the best way to do it.  In particular, I suspect you want getClientBoundingRect rather than querying the individual rects.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+'use strict';
+/* global ApplicationSource */
+/* global BookmarkSource */
+/* global CollectionSource */
```
Are we doing collections and bookmarks for this version of the homescreen, or can these be cut?

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+   * @param {Object} list [DOM container that contains pin elements]
+   */
+  function PinAppNavigation(list) {
+    if (!list && typeof list !== 'object') {
```
Just use !list here.
As the code stands it is buggy. If list is null, the if clause won't run because typeof null === 'object'

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+   */
+  function PinAppNavigation(list) {
+    if (!list && typeof list !== 'object') {
+      console.log('List of pin apps must be an object');
```
Seems more like a console.error() case here.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+  function PinAppNavigation(list) {
+    if (!list && typeof list !== 'object') {
+      console.log('List of pin apps must be an object');
+      return;
```
constructors typically should not return values. The appropriate thing to do here is to throw an exception.  Or just assume that all the call sites will get do the right thing and pass an object and remove the error checking.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+    }
+
+    this.elemList = list;
+    window.removeEventListener('keydown', this);
```
The this object was just created. It is impossible that it is already registered as an event handler. So no need to remove it.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+    elemList: null,
+    prevVerticalKey: null,
+
+    middleElem: false,
```
It is almost never right to define properties like these in the prototype. Put methods in the prototype, but define these properties in the constructor, please.  Or, if you've got a very good reason why these should be in the prototype, add a comment to explain that reason.

davidflanagan replied:
And when you define properties, I find it very helpful if you have a comment saying what each one is for.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+
+
+    /**
+     * This function reset all pinns to default values.
```
I don't understand this comment.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+     * @return {[nothing]} [nothing]
+     */
+    reset: function reset() {
+      MoreAppsNavigation.clearNavigationEvents();
```
This seems bad to have this reset function call a reset function in a different module.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+          return 1;
+        }
+        return elem1.dataset.index - elem2.dataset.index;
+      });
```
Why sort these on reset instead of when you initialize this.elements or add items to it.  Does this array ever get unsorted?

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+
+      for (var elemIndex in this.elements) {
+        this.elemList.appendChild(this.elements[elemIndex]);
+      }
```
I don't get why you're doing this here. If these are not already inserted, a reset() function seems a weird place to insert them.  If they are already inserted, why do they need to be re-inserted?

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+
+      this.selectedElemIndex = 0;
+
+      for (var elemIndex in this.elements) {
```
for/in on an array always seems wrong to me.  I'd prefer a forEach() or the new ES6 for/of loop

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+
+      this.elements[this.selectedElemIndex].classList.add('selected');
+      this.elements[this.selectedElemIndex].classList.add('middle');
+      this.elemList.removeAttribute('style');
```
Working with individual styles is something to avoid when you can help it. I haven't seen what styles you are setting, but I'd rather have you doing removing classes than removing inline styles.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+      this.elements[this.selectedElemIndex].classList.add('selected');
+      this.elements[this.selectedElemIndex].classList.add('middle');
+      this.elemList.removeAttribute('style');
+      this.middleElem = true;
```
Should this be named hasMiddleElement, maybe?  If you had comment this property when you defined it, I would know what it is for.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+      }
+
+      window.removeEventListener('keydown', this);
+      window.addEventListener('keydown', this);
```
I don't get the remove and add.  If the keydown handler has already been added, adding it again will be ignored, so I think it is safe to just leave out the removeEventListener()

davidflanagan replied:
But I don't think it should be necessary to call addEventListener here either. It indicates that you really don't know what state your object is in when this method is called.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+
+      this.elements[this.selectedElemIndex].classList.add('selected');
+        this.elements[this.selectedElemIndex].classList.add('middle');
+        this.middleElem = true;
```
indentation is off on these last two lines.
I can't understand what this function is for. If we're querying a new set of elements, how can selectedElemIndex stay the same?

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+     */
+    handleEvent: function(e) {
+
+      function cycleElements(index) {
```
The argument is the number of elements to cycle, not the index of an element, right? So should you call it n instead of index?

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+        for (var i = 0; i < removed.length; i++) {
+
+          this.elemList.insertBefore(removed[i], this.elements[0 + i]);
+          this.elements.splice(0 + i, 0, removed[i]);
```
What is the 0+i here for? Why not just i?

Delete the extra blank lines and add a comment explaining what this function does.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+      };
+
+      if (this.preventDef) {
+        e.preventDefault();
```
You really can't prevent default until you've checked the keycode and know you're going to handle the key. This will break the volume keys, for example.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+        if (this.middleElem) {
+          document.querySelector('.middle').classList.remove('middle');
+          this.middleElem = false;
+        }
```
If middleElem is a flag, maybe you could just store the middle element in it or null. Then you wouldn't have to do the queryElement()

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+
+        if (toRemoveOutFocus) {
+          toRemoveOutFocus.classList.remove('out-focus');
+        }
```
You had all this code in the reset method. I don't understand yet what it is for, but consider a utility function so you don't have to repeat it.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+
+      if (this.preventDef){
+        e.preventDefault();
+      }
```
You already did this just above.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+      switch(e.key) {
+        case 'ArrowUp':
+
+          preVerticalNavigation.call(this);
```
awkward. Maybe make this into a method so you don't have to do the .call(this)?

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+        case 'ArrowUp':
+
+          preVerticalNavigation.call(this);
+          document.getElementById('clock').style.opacity = 0;
```
don't set the style directly.  And I think you need some kind of hideClock() method or something more generic for this.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+          this.prevVerticalKey = 'up';
+
+
+          break;
```
That is a long case block. You should break it out into a function or method.

Also, the code is pretty much incomprehensible. This is some of the most important code for this homescreen app. It needs to be clear and maintainable.  I suspect it needs to be completely rewritten. But if really needs to be like this, then document it carefully and explain why you're cycling elements and cloning nodes and modifying the position of the container element.  It may be helpful to refer to the UX spec in your comments.

Does the UX spec specify that when the user reaches the end of the list the we cycle back around to the top?  If the code is that complex because of that requirement, consider implementing something simpler to start without that cycling around feature.

I've got the same genral comment for the arrowdown case below, of course.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+        case 'ArrowDown':
+
+          preVerticalNavigation.call(this);
+          document.getElementById('clock').style.opacity = 0;
```
When does the clock become visible again?

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+          this.elements[this.selectedElemIndex].classList.remove('selected');
+          if (this.selectedElemIndex == (this.elements.length - 3)) {
+            if (!this.elemList.dataset.scrolldown) {
+              this.elemList.style.top = parseFloat(this.elemList.style.top) - 6 + 'rem';
```
I don't know why you're doing this, but I wonder if a CSS transform:translate() would be better than messing with style.top like this.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+
+        case 'Accept':
+
+          window.removeEventListener('keydown', this);
```
It doesn't seem like you should have to do this.  If you launch an app, then that app will get focus and we'll lose it, so the handler can remain registered.  In the case where you're just showing the more apps, maybe you can just check a flag to make sure we're not showing more apps before handling the key event here...  It seems like your code would be simpler if you didn't have to keep adding and removing this handler.

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+     * @param  {string} selector [selector that the navigation can take]
+     * @return {[nothing]}          [nothing]
+     */
+    set points_selector(selector) {
```
Wouldn't it be simpler to just pass the selector to the constructor along with the container element?

davidflanagan commented on apps/verticalhome/js/pin_app_navigation.js:
```diff
@@ -0,0 +1,294 @@
+     * @param  {[boolean]} value [if true  preventDefault() function will be called there]
+     * @return {[nothing]}       [nothing]
+     */
+    set prevent(value) {
```
I don't see this being used anywhere. And I can't imagine why it is needed. If you handle a key, call preventDefault on it. Otherwise don't. What am I missing?

davidflanagan commented on apps/verticalhome/js/pin_apps.js:
```diff
@@ -0,0 +1,120 @@
+    pinElem.className = 'pin-app-item';
+    pinElem.setAttribute('data-index', this.index);
+    pinElem.innerHTML = "<img class='pin-app-icon'><br></img>" +
+            "<span class='title'></span>";
```
A <br> inside of an <img> doesn't make sense, does it? This use of innerHTML is maybe okay because it is a string literal. But better to just create the img and span elements explicitly. Then you won't have to query them below.

davidflanagan commented on apps/verticalhome/js/pin_apps.js:
```diff
@@ -0,0 +1,120 @@
+            "<span class='title'></span>";
+    var pinList = document.getElementById('pin-apps-list');
+    var moreAppsLi = document.getElementById('moreApps');
+    pinList.insertBefore(pinElem, moreAppsLi);
```
Is data-index meant to match the index of the element in the pinList.childElements[] array? If so, you probably don't need to pass the index and can calculate it....

davidflanagan commented on apps/verticalhome/js/pin_apps.js:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  PinAppItem.prototype = {
+    entry: null,
```
this should probably not be on the prototype

davidflanagan commented on apps/verticalhome/js/pin_apps.js:
```diff
@@ -0,0 +1,120 @@
+  PinAppItem.prototype = {
+    entry: null,
+
+    bindEntry: function(entry) {
```
could you combine this with the constructor and pass the entry object to the constructor instead of having this separate method?  Also should entry be called something more specific like manifest?

davidflanagan commented on apps/verticalhome/js/pin_apps.js:
```diff
@@ -0,0 +1,120 @@
+    },
+
+    render: function() {
+      this.element.removeEventListener("click", this);
```
Do you really need this? Are you ever going to change the manifest for one of these objects? Why not just register the handler once and leave it there?

davidflanagan commented on apps/verticalhome/js/pin_apps.js:
```diff
@@ -0,0 +1,120 @@
+        }
+      } else {
+          this.icon.style.visibility = "hidden";
+          this.title.style.visibility = "hidden";
```
Indentation is wrong here. Do you ever actually anticipate having one of these things hidden because it has no target app?  Isn't that going to mess up all the navigation code?

davidflanagan commented on apps/verticalhome/js/pin_apps.js:
```diff
@@ -0,0 +1,120 @@
+
+    setEntry: function(entry) {
+      this.bindEntry(entry);
+    },
```
two methods that do exactly the same thing?

davidflanagan commented on apps/verticalhome/js/pin_apps.js:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  PinAppManager.prototype = {
+    items: [],
```
This is a huge bug. If you ever created more than one of these (for writing tests, perhaps) the instances would end up sharing the same items array!  put this in the constructor.

davidflanagan commented on apps/verticalhome/js/pin_apps.js:
```diff
@@ -0,0 +1,120 @@
+    items: [],
+
+    init: function () {
+      this.onLoadSettings();
```
silly to just call another function.  And onLoadSettings() is a bad name because it is not any kind of callback. Just put the contents of that function here in this one.

davidflanagan commented on apps/verticalhome/js/pin_apps.js:
```diff
@@ -0,0 +1,120 @@
+    },
+
+    handleEvent: function(e) {
+    }
```
empty and unused. delete it?

davidflanagan commented on apps/verticalhome/js/pin_apps_default.json:
```diff
@@ -0,0 +1,33 @@
+{"pin_apps_default":
+[
+  {"entry_point": "dialer",
+    "name": "Phone",
```
You can't just hardcode english names like this. The homescreen needs to read a localized name for each app from the app's manifest file.

Localization is important. We've got to support it in this homescreen. If you can't fix this now, we need to at least file a followup bug to be sure you come back to this and do it properly.

davidflanagan commented on apps/verticalhome/js/pin_apps_default.json:
```diff
@@ -0,0 +1,33 @@
+  {"entry_point": "dialer",
+    "name": "Phone",
+    "manifestURL": "app://communications.gaiamobile.org/manifest.webapp",
+    "icon":   "app://communications.gaiamobile.org/dialer/style/icons/dialer_84.png",
```
Does this work? Can the homescreen just read these images cross-origin like this?  As a general thing, you can't count on every app having a hardcoded 84 size icon. I think there is utility code in the existing homescreen for finding the most appropriate size icon for a given app. You'll probably need to use something like that.

davidflanagan commented on apps/verticalhome/js/pin_apps_default.json:
```diff
@@ -0,0 +1,33 @@
+    "name": "Phone",
+    "manifestURL": "app://communications.gaiamobile.org/manifest.webapp",
+    "icon":   "app://communications.gaiamobile.org/dialer/style/icons/dialer_84.png",
+    "index": 0
```
Seems like this index is unnecessary since these are array elements that have an index implicitly.

davidflanagan commented on apps/verticalhome/manifest.webapp:
```diff
@@ -15,6 +15,7 @@
     "storage": {},
     "webapps-manage": {},
     "systemXHR": {},
+    "alarms": {},
```
What is this going to be for?  I haven't see any code using this permission yet.  Are you sure you want to add this now?

davidflanagan commented on apps/verticalhome/js/statusbar.js:
```diff
@@ -11,7 +11,7 @@
 
   function StatusBar() {
-    this.threshold = document.getElementById('search').clientHeight;
+//    this.threshold = document.getElementById('search').clientHeight;
```
There is probalby a lot more code here that you can remove...

davidflanagan commented on apps/verticalhome/js/stores/item.js:
```diff
@@ -360,15 +474,6 @@
-            addIfUnique.call(this, bookmark);
-          } else if (thisItem.type === 'collection') {
-            var collection = new GaiaGrid.Collection(thisItem);
-            addIfUnique.call(this, collection);
```
Looks like you can remove the Divider, Bookmark and Collection classes, given that you're removed this code.

davidflanagan commented on apps/verticalhome/style/css/pin_apps.css:
```diff
@@ -0,0 +1,102 @@
+
+
+
+
```
please try to avoid this kind of repeated blank lines

davidflanagan commented on apps/verticalhome/js/stores/item.js:
```diff
@@ -194,9 +195,13 @@
           objectSV.createIndex('indexSV', 'indexSV', { unique: true });
+
+          var pinAppStore = db.createObjectStore(DB_PIN_APP_STORE, { keyPath: 'index'});
+          pinAppStore.createIndex('index', 'index', { unique: true });
```
Maybe I'm misremembering my indexedDB, but I didn't think you had to create an index for the primary index of an object store....  I see the existing code does that, but I thought that this keyPath thing would mean you didn't have to create the index.  Though I would have thought that your keypath would be the manifest URL or something...

Maybe I'm just not clear on what you'r doing here.

davidflanagan replied:
It seems like a bad idea to create an index with the same name as the key that is being indexed.  I can't tell if this index is ever actually used. (Maybe loadTable uses it?)

davidflanagan commented on apps/verticalhome/js/stores/item.js:
```diff
@@ -244,6 +255,14 @@
 
+
+    /**
+     * A list that contains applications displaies on the main screen
```
typo

davidflanagan commented on apps/verticalhome/js/stores/item.js:
```diff
@@ -270,6 +307,83 @@
+                    var DBIndex = objFromDB.index;
+                    objFromDB = object;
+                    objFromDB.index = DBIndex;
+                    store.put(objFromDB);
```
This code is a mess. I think you just want:
```
if (req.result) {
  object.index = req.result.idex;
}
store.put(object);
```

davidflanagan commented on apps/verticalhome/js/stores/item.js:
```diff
@@ -270,6 +307,83 @@
+                  }
+                }
+              },
+              callback);
```
So this function allows you to blow away existing database entries by just re-using an existing index, it seems.
Wouldn't it be better to use the manifestURL (which is the unique identifier for apps) as the key for this object store? Especially since you're also creating an index for the index property?

MariaKolesina committed:
Bug 1217816 - [RedSquare] Homescreen - More Apps view

Changed view of "More Apps" and added touchless navigation

ruvsmirnov committed:
Bug 1210304 - David's comments related to clock were fixed..

Clock functionality was transfered (installing localization and handling timeformatchange event) from "app.js" file to "clock.js" file.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b23360a1dcce5bf16b232bdd89e741ab6670d80b

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=31a4a1c021e26756313fc09bffd4142d2b3157b7

MariaKolesina committed:
Bug 1210304 - fixed some David's comments

1) Rather than using the hidden attribute here, it is probably better to set and remove a CSS class that will show the more apps screen on top of the main screen. That way you can specify CSS animations for the transition, which is what our UX designers have in mind, I think.
Done.
2) just use the hidden attribute with no value. No need for the =""
In accordance with previous comment changed to class.
3) If you ever have to use proto your code is almost certainly wrong. Accessing proto is also a big performance hit, since the code can't be optimized by the JS interpreter. You'll have to have a very good reason for proto and a long comment explaining the reason, or I will not land the patch.
Removed proto.
4) To be parallel with the function above, I'd suggest you call this hideMoreApps()
Renamed.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4b3a9cbb1a60737140688471defe2f843513909a

Commit:
Bug 1210304 - Fix some David's comments about more apps navigation

More apps navigation has been changed
Comments for all functions in more_apps_navigation.js have been added.

ruvsmirnov committed:
Bug 1210304 - Add localization of Main screen's icons

**Issues that was done:**
  * Add localization of Main screen's icons
  * Fix lint errors for verticalhomescreen

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5ea29984d0aae42b85a8d473a1136b54a3a86198

MariaKolesina committed:
Bug 1210304 - Unused settings has been removed

David's comment:
"Just removing this one select element leave us with an empty panel, doesn't it? I think you should probably remove this entire file and the references to it, right?
And of course you should also remove the homescreen code that reads this grid.layout.cols setting, whereever that code is."
Removed homescreen setting for changing number of columns

MariaKolesina committed:
Bug 1210304 - David's comment about number of columns in JSON

David's comment:
"Given that the number of columns will not be configurable, I think that it doesn't need to be in this JSON file anymore. Maybe find the code that reads this file and just define a const for the number of columns."
Removed number of columns from default-homescreens.json. Define a const for the number of columns in configurator.js

Commit:
Merge pull request #17 from MariaKolesina/bug-homescreen-fixcomments-removesettings

Bug 1210304 - Unused settings has been removed

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9001cc456b4432ee593bfced32efc332cb96db4e

Commit:
Merge pull request #18 from MariaKolesina/bug-homescreen-fixcomments-columns

Bug 1210304 - David's comment about number of columns in JSON

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9abb74ef6ca01f26294da56c96bc48b19fe99abf

Commit:
Merge pull request #24 from MariaKolesina/bug-1210304-homescreen-moreappsnavigation

Bug 1210304 - Fix some David's comments about more apps navigation

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=43646f5d2fad31d8f23fcf555dfa96fba379ac65

Commit:
Merge pull request #23 from ruvsmirnov/homescreen-pinApp-localization

Bug 1210304 - Add localization of Main screen's icons

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=aee8230a66ee93c327d6d10ffcebd6ae6571367a

Commit:
Bug 1210304 - Fix some David's comments

Renamed app_manager.js to more_apps_manager.js
Removed some unused code (search, statusbar)
Fix comment in more_apps_navigation.js about getBoundingClientRect

ruvsmirnov committed:
Bug 1210304 - David's comments by homescreen

**David's comments mostly related to pin app navigation:**
  * Fixed minor bugs
  * Removed not used strings
  * Deleted free spaces
  * Initialization of veriables moved to constructor
  * Everywhere changed .call(this) to self=this
  * Changed some logic of pinned apps navigation

Commit:
Bug 1210304 - fix David's comment about unused code

In verticalhome app:
Collections has been removed
Bookmarks has been removed
Divider has been removed

Commit:
Bug 1210304 - Renamed pinApp to pinnedApp

Commit:
Merge pull request #30 from MariaKolesina/bug-1210304-homescreen-fixcomments

Bug 1210304 - Fix some David's comments

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c9673651d31878c06c3906d4b8b31970138ebbc9

Commit:
Merge pull request #33 from MariaKolesina/bug-1210304-homescreen-fixcomments-pinned

Bug 1210304 - Renamed pinApp to pinnedApp

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d7a5561e967df38ee90c8d374dd6cd322e5e248f

Commit:
Merge pull request #32 from MariaKolesina/bug-1210304-fixcomments-removedcode

Bug 1210304 - fix David's comment about unused code

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a4903fbca92482d001008b6bae1fe6771095a377

Commit:
Merge pull request #29 from ruvsmirnov/homescreen-Davids-comments

Bug 1210304 - David's comments by homescreen

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=cd4708a014e760c3b7aa77af5c089e7454f79cb6

TatianaMigacheva commented:
This PR has been closed because RedSquare project has been canceled. All changes which were done by Bug#1210304 will be attached as Bugzilla patch.
